### PR TITLE
Improve call, and usage in pallet utility

### PR DESCRIPTION
**Pull request by thiolliere**
this decrease the length of call to less than 200 (184 exactly) bytes.
and improve the usage in utility.

companion: https://github.com/paritytech/polkadot/pull/3522

### CODE REVIEW

**Comment by thiolliere** [comment]
@jacogr I have 2 notes:
* Some calls now have a `Box<...>` for their type instead of just the type directly, the encoding is same as the inner type. Does this needs consideration from UI ?
* batch and batch_all now have a strong limit on the number of batched calls, the limit can be found on metadata, this can also needs consideration from UI.

any remark is welcomed

**Comment by jacogr** [comment]
Box wrappings are ignored and stripped. (It has no meaning outside the Rust code)

If there is a batch limit on constants, can be read.

File: bin/node/runtime/src/lib.rs [hunk 1]
```diff
@@ -1626,4 +1626,14 @@ mod tests {
...
+
+	#[test]
+	fn call_size() {
+		assert!(
+			core::mem::size_of::<Call>() <= 200,
+			"size of Call is more than 200: some calls have too big arguments, use Box to reduce the
+			size of Call.
+			If the limit is too strong, maybe consider increase the limit to 300.",
+		);
+	}
```
Comment (ordian) [inline bin/node/runtime/src/lib.rs:1638]: have you considered making a `static_assertion!` instead?

File: frame/utility/src/lib.rs [hunk 2]
```diff
@@ -110,13 +110,33 @@ pub mod pallet {
 		BatchCompleted,
 	}
 
+	#[pallet::extra_constants]
+	impl<T: Config> Pallet<T> {
+		/// The limit on the number of batched calls.
+		fn batched_calls_limit() -> u32 {
+			let allocator_limit = 33554432; // 32MiB
```
Comment (shawntabrizi) [inline frame/utility/src/lib.rs:117]: should be available now thanks to other pr?

File: bin/node/runtime/src/lib.rs [hunk 3]
```diff
@@ -1626,4 +1626,14 @@ mod tests {
 
 		is_submit_signed_transaction::<Runtime>();
 	}
+
+	#[test]
+	fn call_size() {
+		assert!(
+			core::mem::size_of::<Call>() <= 200,
```
Comment (shawntabrizi) [inline bin/node/runtime/src/lib.rs:1633]: 200 is an arbitrary number?

**Review by shawntabrizi — APPROVED** [review]
looks okay to me, probably need more input from others

File: frame/utility/src/lib.rs (hunk 2 above)
Comment (thiolliere) [reply frame/utility/src/lib.rs:117]: yes I'll update once https://github.com/paritytech/substrate/pull/9393 is merged

File: bin/node/runtime/src/lib.rs (hunk 3 above)
Comment (thiolliere) [reply bin/node/runtime/src/lib.rs:1633]: yes IIRC it is the same number as clippy lint

File: bin/node/runtime/src/lib.rs (hunk 1 above)
Comment (thiolliere) [reply bin/node/runtime/src/lib.rs:1638]: I gave a try but I think the error message is better with a regular assert.
But all in all this assert is not really important there is no important issue with using a big call enum

```
error[E0080]: evaluation of constant value failed
    --> bin/node/runtime/src/lib.rs:1632:3
     |
1632 | /         static_assertions::const_assert!(
1633 | |             core::mem::size_of::<Call>() <= 20,
1634 | |         );
     | |__________^ attempt to compute `0_usize - 1_usize`, which would overflow
     |
     = note: this error originates in the macro `static_assertions::const_assert` (in Nightly builds, run with -Z macro-backtrace for more info)

```

File: bin/node/runtime/src/lib.rs [hunk 4]
```diff
@@ -1626,4 +1626,14 @@ mod tests {
...
+			"size of Call is more than 200: some calls have too big arguments, use Box to reduce the
```
Comment (bkchr) [inline bin/node/runtime/src/lib.rs:1634]:
```suggestion
			"size of Call is more than 200 bytes: some calls have too big arguments, use Box to reduce the
```

File: bin/node/runtime/src/lib.rs (hunk 1 above)
Comment (bkchr) [reply bin/node/runtime/src/lib.rs:1638]: Until const panics are a thing, static assertions are rather useless.

**Review by bkchr — APPROVED** [review]

File: bin/node/runtime/src/lib.rs [hunk 5]
```diff
@@ -1626,4 +1626,14 @@ mod tests {
 
 		is_submit_signed_transaction::<Runtime>();
 	}
+
+	#[test]
+	fn call_size() {
```
Comment (kianenigma) [inline bin/node/runtime/src/lib.rs:1631]: You should add the same to polkadot/kusama/westend runtimes as well to ensure their size also does not change over time.

File: bin/node/runtime/src/lib.rs (hunk 3 above)
Comment (kianenigma) [reply bin/node/runtime/src/lib.rs:1633]: what clippy lint?

**Review by kianenigma — APPROVED** [review]
LGTM + https://github.com/paritytech/substrate/pull/9418/files#r675454591

File: bin/node/runtime/src/lib.rs (hunk 3 above)
Comment (thiolliere) [reply bin/node/runtime/src/lib.rs:1633]: this one https://rust-lang.github.io/rust-clippy/master/#large_enum_variant

File: bin/node/runtime/src/lib.rs (hunk 5 above)
Comment (thiolliere) [reply bin/node/runtime/src/lib.rs:1631]: yes, I'll do in the companion

Comment (thiolliere) [reply bin/node/runtime/src/lib.rs:1632]: companion should be ready now https://github.com/paritytech/polkadot/pull/3522

File: frame/utility/src/lib.rs (hunk 2 above)
Comment (thiolliere) [reply frame/utility/src/lib.rs]: fixed

**Comment by shawntabrizi** [comment]
if we want to prevent a large memory allocation before decoding, this pr is not good enough. this would only prevent it after decode

the proposed solution from sr labs is to instead make the signature of `batch` take `vec<box<call>>` where now we know each `box<call>` will be 4 bytes, and then the vec should have no problems w/ size before decode.

However, this is potentially a huge performance decrease, especially as the vec size increases

**Comment by thiolliere** [comment]
> the proposed solution from sr labs is to instead make the signature of `batch` take `vec<box<call>>` where now we know each `box<call>` will be 4 bytes, and then the vec should have no problems w/ size before decode.

IIRC it was my idea, it is implemented here https://github.com/paritytech/substrate/pull/9339

**Comment by bkchr** [comment]
> if we want to prevent a large memory allocation before decoding, this pr is not good enough. this would only prevent it after decode

If we can not decode a call, it is invalid anyway. I don't see the reason why we need to work here on better solutions.

**Comment by thiolliere** [comment]
I think this PR improve call by reducing its size, and makes utility batch function limit more explicit for users.

So we can merge it. Otherwise we can close it

**Comment by bkchr** [comment]
I'm for merging this.

**Comment by shawntabrizi** [comment]
okay, then lets go ahead and do that :)

I just wanted to post what was said in my last meeting with the auditors.

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**